can/can.cpp

```diff
@@ -0,0 +1,255 @@
+
+    return CAN_ClassSettingStatus::NON_ERROR;
+}
+
+CAN_ClassSettingStatus Can::setTwoTypePathIdGroup(uint32_t minId1, uint32_t maxId1,
+
+uint32_t minId2, uint32_t maxId2) {
+    if ((filterMode != CAN_FilterMode::PATH_ALL_ID) && (filterMode != CAN_FilterMode::PATH_TWO_TYPE_STD_ID_GROUP)) {
+        return CAN_ClassSettingStatus::WRONG_SET_PATH_ID_FUNCTION_USED;
+    } else if (filterMode == CAN_FilterMode::PATH_ALL_ID) {
+        return CAN_ClassSettingStatus::ALL_ID_PATH;
+    }
+
+    filterId[0] = minId1;
+    minId1 <<= 21;
+    maxId1 <<= 21;
+    filterMask[0] = ((~(minId1 ^ maxId1)) >> 21);
+
+    filterId[1] = minId2;
+    minId2 <<= 21;
+    maxId2 <<= 21;
+    filterMask[1] = ((~(minId2 ^ maxId2)) >> 21);
+
+    filterConfig.FilterIdHigh = filterId[0] << 5;
+    filterConfig.FilterIdLow = filterId[1] << 5;
+
+    filterConfig.FilterMaskIdHigh = filterMask[0] << 5;
+    filterConfig.FilterMaskIdLow = filterMask[1] << 5;
+
+    return CAN_ClassSettingStatus::NON_ERROR;
+}
+
+CAN_ClassSettingStatus Can::setTwoTypePathId(uint32_t idType1, uint32_t id1, uint32_t idType2, uint32_t id2) {
+    if ((filterMode != CAN_FilterMode::PATH_ALL_ID) && (filterMode != CAN_FilterMode::PATH_TWO_TYPE_STD_OR_EXT_ID)) {
+        return CAN_ClassSettingStatus::WRONG_SET_PATH_ID_FUNCTION_USED;
+    } else if (filterMode == CAN_FilterMode::PATH_ALL_ID) {
+        return CAN_ClassSettingStatus::ALL_ID_PATH;
+    }
+
+    if (idType1 == CAN_ID_STD) {
+        filterId[0] = id1 << 21;
+    } else if (idType1 == CAN_ID_EXT) {
+        filterId[0] = (id1 << 3) | 0x4;
+    }
+
+    if (idType2 == CAN_ID_STD) {
+        filterId[1] = id2 << 21;
+    } else if (idType2 == CAN_ID_EXT) {
+        filterId[1] = (id2 << 3) | 0x4;
+    }
+
+    filterConfig.FilterIdHigh = filterId[0] >> 16;
+    filterConfig.FilterIdLow = filterId[0];
+
+    filterConfig.FilterMaskIdHigh = filterId[1] >> 16;
+    filterConfig.FilterMaskIdLow = filterId[1];
+
+    return CAN_ClassSettingStatus::NON_ERROR;
+}
+
+CAN_ClassSettingStatus Can::setOneTypePathIdGroup(uint32_t idType, uint32_t minId, uint32_t maxId) {
+    if ((filterMode != CAN_FilterMode::PATH_ALL_ID) && (filterMode != CAN_FilterMode::PATH_ONE_TYPE_STD_OR_EXT_ID_GROUP)) {
+        return CAN_ClassSettingStatus::WRONG_SET_PATH_ID_FUNCTION_USED;
+    } else if (filterMode == CAN_FilterMode::PATH_ALL_ID) {
+        return CAN_ClassSettingStatus::ALL_ID_PATH;
+    }
+
+    std::array<uint8_t, 2> bitShift;
+    if (idType == CAN_ID_STD) {
+        bitShift[0] = 21;
+    } else if (idType == CAN_ID_EXT) {
+        bitShift[0] = 3;
+    }
+
+    filterId[0] = minId << bitShift[0];
+
+    minId <<= 21;
+    maxId <<= 21;
+    filterMask[0] = ((~(minId ^ maxId)) >> 21);
+    filterMask[0] <<= bitShift[0];
+
+    filterConfig.FilterIdHigh = filterId[0] >> 16;
+    filterConfig.FilterIdLow = filterId[0];
+
+    filterConfig.FilterMaskIdHigh = filterMask[0] >> 16;
+    filterConfig.FilterMaskIdLow = filterMask[0];
```

Comment:
```suggestion
    uint8_t bitShift;
    if (idType == CAN_ID_STD) {
        bitShift = 21;
    } else if (idType == CAN_ID_EXT) {
        bitShift = 3;
    }

    filterId[0] = minId << bitShift[0];

    minId <<= 21;
    maxId <<= 21;
    filterMask[0] = ((~(minId ^ maxId)) >> 21);
    filterMask[0] <<= bitShift;

    filterConfig.FilterIdHigh = filterId[0] >> 16;
    filterConfig.FilterIdLow = filterId[0];

    filterConfig.FilterMaskIdHigh = filterMask[0] >> 16;
    filterConfig.FilterMaskIdLow = filterMask[0];
```

[MAY] bitShift[1] を使っていないので array を書き換えられそうです

Reply:
bac9fe61810e1a8672e7f552cda80e15a306e74c で対応